repair/repair.cc

```diff
@@ -1959,23 +1985,66 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get0();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring ks_source_dc;
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (!source_dc_live_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    ssize_t rf = nt_strat.get_replication_factor(source_dc);
+                    size_t nr_racks = source_dc_racks.size();
+                    size_t max_replicas_per_rack = (rf + nr_racks - 1) / nr_racks;
+                    size_t possibly_lost_replicas = 0;
+                    for (const auto& [rack, nr] : lost_nodes_per_rack) {
+                        possibly_lost_replicas += std::min(nr, max_replicas_per_rack);
+                    }
+                    // Can we ensure a quorum in source_dc?
+                    if (possibly_lost_replicas < rf / 2) {
+                        ks_source_dc = source_dc;
+                    } else if (reason == streaming::stream_reason::rebuild) {
+                        if (rf == 0) {
+                            throw std::runtime_error(fmt::format("Cannot rebuild with source_dc={}: Replication factor in this data center is 0."
+                                    " Use a different source_dc, or no source_dc option to consider the whole cluster for repair", source_dc));
+                        }
+                        rlogger.warn("{}: keyspace={}: Cannot ensure quorum in source_dc={}: possibly lost {} replicas out of {}: rebuilding using source_dc anyway",
+                                op, keyspace_name, source_dc, possibly_lost_replicas, rf);
+                        ks_source_dc = source_dc;
+                    } else {
+                        rlogger.warn("{}: keyspace={}: Cannot ensure quorum in source_dc={}: possibly lost {} replicas out of {}: Falling back to full-cluster repair",
```

Comment:
It is better to say 

Failing back to replace / rebuild with nodes in all data centers.  

The full-cluster repair would confuse user.

Reply:
okay, no problem

Reply:
See v5